test/test_helper.exs

```diff
@@ -1 +1,90 @@
+defmodule EventCatcher do
+  @moduledoc """
+  A listener that stashes received events.
+  """
+  use GenServer
+
+  def start_link(opts), do: GenServer.start_link(__MODULE__, [], opts)
+  def events(server), do: GenServer.call(server, :events)
+
+  def handle_call(:events, _from, events), do: {:reply, events, events}
+  def handle_info({:sse, event}, events), do: {:noreply, [event | events]}
+end
+
+defmodule FakeMarathon do
+  @moduledoc """
+  A fake Marathon API that can stream events.
+  """
+  use GenServer
+
+  defmodule HelloHandler do
+    def init(req, state) do
+      new_req = :cowboy_req.reply(
+        200,
+        %{"content-type" => "text/html"}, "hello", req)
+      {:ok, new_req, state}
+    end
+  end
+
+  defmodule SSEHandler do
+    def init(req, state) do
+      new_req = :cowboy_req.stream_reply(
+        200, %{"content-type" => "text/event-stream"}, req)
+      FakeMarathon.sse_stream(:fake_marathon, self())
+      {:cowboy_loop, new_req, state}
+    end
+
+    def info(:keepalive, req, state) do
+      :cowboy_req.stream_body("\r\n", :nofin, req)
+      {:ok, req, state}
+    end
+
+    def info({:event, event, data}, req, state) do
+      ev = "event: #{event}\r\ndata: #{data}\r\n\r\n"
+      :cowboy_req.stream_body(ev, :nofin, req)
+      {:ok, req, state}
+    end
+  end
+
+  defmodule State do
+    defstruct port: nil, sse_streams: []
+  end
+
+  ## Client
+
+  def start_link(_) do
+    GenServer.start_link(__MODULE__, :ok, name: :fake_marathon)
+  end
+  def port(fm), do: GenServer.call(fm, :port)
+  def base_url(fm), do: "http://localhost:#{port(fm)}"
+  def event(fm, event, data), do: GenServer.call(fm, {:event, event, data})
+  def sse_stream(fm, pid), do: GenServer.call(fm, {:sse_stream, pid})
+
+  ## Callbacks
+
+  def init(:ok) do
+    handlers = [
+      {"/", HelloHandler, []},
+      {"/v2/events", SSEHandler, []},
+    ]
+    dispatch = :cowboy_router.compile([{:_, handlers}])
+    {:ok, listener} = :cowboy.start_clear(
+      :fm_listener, [], %{env: %{dispatch: dispatch}})
```

Comment:
I'm not sure I have the energy to read through the cowboy docs and understand this completely, but it _seems_ reasonable.

Reply:
I'm not any kind of expert on `cowboy`, but this comes pretty much directly from the docs with some syntax translation (aided by looking at an Elixir-based example repo for an older version which used obsolete APIs).